### PR TITLE
[cmake] provide internal depends builds with build_type

### PR DESCRIPTION
**Pull request by fuzzard**
## Description
Provide Internal depends targets that are cmake projects themselves with the build type (Debug/release/relwithdebinfo etc)

## Motivation and context
Allow ability to build dependencies with different build types. 
~~CrossGUID, FlatBuffers are also cmake projects that could be done, but i was unsure of the need for debug versions of them. If we want them done for uniformity, i can get them added to this, let me know.~~
~~The other "cmake" based internal dep is FFMPEG, again, ive left it untouched, as it just needs one big overhaul.~~

The discussion may be whether we need to provide non debug builds of these libs for most devs? Should we just stick with defaulting to release for everything? Happy to close out the PR if thats the decision.
~~Do we introduce a variable to override individually for the internal build? eg INTERNAL_SPDLOG_BUILDTYPE, that relies on ENABLE_INTERNAL_SPDLOG to activate?~~

Due to #21052 this is simplified to just automatically be done for any cmake based internal dep.
This now allows for overrides of the internal dep build type to be given by the user (eg -DFFMPEG_BUILD_TYPE=Relwithdebinfo -DCMAKE_BUILD_TYPE=Release) will have the project build as release, but FFMPEG will build as Relwithdebinfo.

## How has this been tested?
Makefile build spdlog/fmt as Debug and Release
Multi-config generator (Xcode) schemes build spdlog/fmt with the chosen build type (Debug/release/relwithdebinfo)

## What is the effect on users?
N/A

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by garbear** [comment]
Flatbuffers native should always be release I think. Internal flatbuffers might be headers only, but I like the idea of uniformity.

**Comment by fuzzard** [comment]
After the merge of #21052 ive reworked this. Uniformity is now just a given for any cmake based dep build.

Ive allowed an override, so an individual module can set its build_type regardless of the project. This covers @garbear 's comment about Flatbuffers being always release (build tool is native, no need for target build_type). Gtest is in the same boat being a native tool, theres no real reason to build as target type (eg Debug)

File: cmake/modules/FindFmt.cmake [hunk 1]
```diff
@@ -33,6 +33,15 @@ if(ENABLE_INTERNAL_FMT)
     set(EXTRA_ARGS "-DCMAKE_OSX_ARCHITECTURES=${CMAKE_OSX_ARCHITECTURES}")
   endif()
 
+  # single config generator (ie Make)
+  if(CMAKE_BUILD_TYPE)
+    set(BUILD_TYPE ${CMAKE_BUILD_TYPE})
+  else()
+    # Multi-config generators (eg VS,Xcode, Ninja) will not have CMAKE_BUILD_TYPE
```
[the remark this answers is not in the document]
Comment (fuzzard) [reply cmake/modules/FindFmt.cmake]: Have updated the comments

**Review by wsnipex — APPROVED** [review]
looks good. Maybe update docs ;)

**Comment by fuzzard** [comment]
There was a discussion around docs and our cmake options in slack a day or two ago. For now, i wont do anything in this PR, but i'll start to draft up some docs for our cmake build options, and let the bike shedding happen on that PR.

**Comment by wsnipex** [comment]
looks like there is no unified document, so at least here:
https://github.com/xbmc/xbmc/blob/master/docs/README.Linux.md#32-enable-internal-dependencies

up to you if you want to add it to darwin docs

File: cmake/scripts/common/ModuleHelpers.cmake [hunk 2]
```diff
@@ -136,6 +136,21 @@ macro(BUILD_DEP_TARGET)
...
+    # Set build type for dep build.
+    # if MODULE has set a manual build type, use it, otherwise use project build type
+    if(${MODULE}_BUILD_TYPE)
+      list(APPEND CMAKE_ARGS "-DCMAKE_BUILD_TYPE=${${MODULE}_BUILD_TYPE}")
+    else()
+      # single config generator (ie Make, Ninja)
+      if(CMAKE_BUILD_TYPE)
+        list(APPEND CMAKE_ARGS "-DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}")
+      else()
+        # Multi-config generators (eg VS,Xcode, Ninja Multi-Config) will not have CMAKE_BUILD_TYPE
```
Comment (garbear) [inline cmake/scripts/common/ModuleHelpers.cmake:149]: If you're able to care about a stray grammar issue, `,` deserves a space after

**Review by garbear — APPROVED** [review]
Looks like the above comments have been addressed, I have nothing more to add.

As your goal is to add a bunch of dependencies to internal-build-ability, maybe just update this comment to say "That's the case with multiple dependencies": https://github.com/xbmc/xbmc/blob/master/docs/README.Linux.md#32-enable-internal-dependencies

**Comment by howie-f** [comment]
building in Debug mode with internal `fmt` enabled compiles to static library `libfmtd.a` now (notice the **d** = _Debug_).
```
-rw-r--r--  1 fhw fhw  1761164 May 24 19:16 libfmtd.a

Scanning dependencies of target kodi
gmake[2]: *** No rule to make target 'build/lib/libfmt.a', needed by 'kodi-x11'.  Stop.
gmake[2]: *** Waiting for unfinished jobs....
[100%] Building CXX object CMakeFiles/kodi.dir/xbmc/platform/posix/main.cpp.o
gmake[1]: *** [CMakeFiles/Makefile2:5255: CMakeFiles/kodi.dir/all] Error 2
gmake: *** [Makefile:160: all] Error 2
```

you don't notice until internal dependencies are cleared and rebuilt.

i suspect it's named this way because of this https://github.com/fmtlib/fmt/blob/d141cdbeb0fb422a3fb7173b285fd38e0d1772dc/CMakeLists.txt#L269-L280

**Comment by fuzzard** [comment]
Should be correct, and it should work at a glance on my phone. I'll have a dig into it when I can.
In the meantime you can do something like 
```
set(FMT_BUILD_TYPE Release)
```

Just before the set(CMAKE_ARGS ) call

**Pull request merged**